### ISSUE TITLE
Report statistics as histogramm

### ISSUE THREAD

**Issue by jlink** [jlink/jqwik#83]
## Testing Problem

In https://github.com/jlink/jqwik/issues/75#issuecomment-575592315 it has been suggested to report statistics as a simple ascii graphic diagram.

## Suggested Solution

Report optionally in current form or as histogramm. Inspired by https://tapoueh.org/blog/2014/02/postgresql-aggregates-and-histograms/.

## Discussion

- How should report format be configured?
- How to go about multi-variable statistics?

**Comment by jlink** [comment]
A first step is made. You can plug in your own statistics report format like that:

```
@Property
@StatisticsReport(format = MyStatisticsFormat.class)
void statisticsWithHandMadeFormat(@ForAll Integer anInt) {
	String range = anInt < 0 ? "negative" : anInt > 0 ? "positive" : "zero";
	Statistics.collect(range);
}

class MyStatisticsFormat implements StatisticsReportFormat {
	@Override
	public List<String> formatReport(List<StatisticsEntry> entries) {
		return entries.stream()
					  .map(e -> String.format("%s: %d", e.name(), e.count()))
					  .collect(Collectors.toList());
	}
}
```

This can be done starting version "1.2.3-SNAPSHOT"

@luvarqpp Feel free to show an implementation for a histogram if you feel like it ;-)

**Comment by jlink** [comment]
Current state of implementation:

```java
class HistogramExamples {

	@Property(generation = GenerationMode.RANDOMIZED)
	@StatisticsReport(format = Histogram.class)
	void integers(@ForAll("gaussians") int aNumber) {
		Statistics.collect(aNumber);
	}

	@Provide
	Arbitrary<Integer> gaussians() {
		return Arbitraries
				   .integers()
				   .between(0, 20)
				   .shrinkTowards(10)
				   .withDistribution(RandomDistribution.gaussian());
	}

}
```

This will show the following output:

```
[HistogramExamples:integers] (1000) statistics = 
       # | label | count | 
    -----|-------|-------|---------------------------------------------------------------------------------
       0 |     0 |    17 | ■■■■■
       1 |     1 |    13 | ■■■■
       2 |     2 |    20 | ■■■■■■
       3 |     3 |     8 | ■■
       4 |     4 |    20 | ■■■■■■
       5 |     5 |    30 | ■■■■■■■■■■
       6 |     6 |    50 | ■■■■■■■■■■■■■■■■■
       7 |     7 |    60 | ■■■■■■■■■■■■■■■■■■■■
       8 |     8 |   103 | ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
       9 |     9 |   105 | ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
      10 |    10 |   233 | ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
      11 |    11 |    80 | ■■■■■■■■■■■■■■■■■■■■■■■■■■■
      12 |    12 |    85 | ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
      13 |    13 |    57 | ■■■■■■■■■■■■■■■■■■■
      14 |    14 |    42 | ■■■■■■■■■■■■■■
      15 |    15 |    34 | ■■■■■■■■■■■
      16 |    16 |    24 | ■■■■■■■■
      17 |    17 |     6 | ■■
      18 |    18 |     5 | ■
      19 |    19 |     1 | 
      20 |    20 |     7 | ■■
```

**Comment by jlink** [comment]
And ideas for a good DSL to group collected values?

**Comment by jlink** [comment]
Number ranges can be shown as histograms:

```
@Property(generation = GenerationMode.RANDOMIZED)
@StatisticsReport(format = NumberRangeHistogram.class)
void integersInRanges(@ForAll @IntRange(min = -1000, max = 1000) int aNumber) {
	Statistics.collect(aNumber);
}
```

Output:
```
[HistogramExamples:integersInRanges] (1000) statistics = 
       # |         label | count | 
    -----|---------------|-------|---------------------------------------------------------------------------------
       0 | [-1000..-900[ |    28 | ■■■■■■■
       1 |  [-900..-800[ |    11 | ■■
       2 |  [-800..-700[ |    12 | ■■■
       3 |  [-700..-600[ |    15 | ■■■
       4 |  [-600..-500[ |    12 | ■■■
       5 |  [-500..-400[ |    17 | ■■■■
       6 |  [-400..-300[ |    21 | ■■■■■
       7 |  [-300..-200[ |    42 | ■■■■■■■■■■
       8 |  [-200..-100[ |    52 | ■■■■■■■■■■■■■
       9 |     [-100..0[ |   275 | ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
      10 |      [0..100[ |   316 | ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
      11 |    [100..200[ |    49 | ■■■■■■■■■■■■
      12 |    [200..300[ |    49 | ■■■■■■■■■■■■
      13 |    [300..400[ |    23 | ■■■■■
      14 |    [400..500[ |    12 | ■■■
      15 |    [500..600[ |    12 | ■■■
      16 |    [600..700[ |     7 | ■
      17 |    [700..800[ |     8 | ■■
      18 |    [800..900[ |    19 | ■■■■
      19 |   [900..1000] |    20 | ■■■■■
```

All feedback welcome!

**Comment by jlink** [comment]
Added user guide entry: http://127.0.0.1:4000/docs/snapshot/user-guide.html#histograms

**Comment by jlink** [comment]
Added user-guide entry: https://jqwik.net/docs/snapshot/user-guide.html#histograms

**Comment by jlink** [comment]
Available in `1.3.0-SNAPSHOT`.
